src/cryptsetup/cryptsetup-generator.c

```diff
@@ -285,21 +285,72 @@ static int print_dependencies(FILE *f, const char* device_path, const char* time
         return 0;
 }
 
-static bool attach_in_initrd(const char *name, const char *options) {
+static bool attach_in_initrd(const char *name, const char *device, const char *uuid, const char *options) {
+        int r;
+
         assert(name);
 
         /* Imply x-initrd.attach in case the volume name is among those defined in the Discoverable Partition
          * Specification for partitions that we require to be mounted during the initrd → host transition,
          * i.e. for the root fs itself, and /usr/. This mirrors similar behaviour in
          * systemd-fstab-generator. */
 
-        return fstab_test_option(options, "x-initrd.attach\0") ||
-                STR_IN_SET(name, "root", "usr");
+        if (fstab_test_option(options, "x-initrd.attach\0") ||
+            STR_IN_SET(name, "root", "usr"))
+                return true;
+
+        /* Also return true if the UUID is specified in rd.luks.uuid= kernel command line parameter */
+        if (uuid && hashmap_contains(arg_disks, uuid))
+                return true;
+        else {
+                _cleanup_strv_free_ char **cmdline = NULL;
+
+                /* UUID is NULL so let's check command line manually */
+                uuid = startswith(device, "UUID=");
+                if (!uuid)
+                        uuid = path_startswith(device, "/dev/disk/by-uuid/");
+                if (!uuid)
+                        uuid = startswith(name, "luks-");
+
+                if (!uuid)
+                        return false;
+
+                r = proc_cmdline_strv(&cmdline);
+                if (r < 0)
+                        return false;
+
+                STRV_FOREACH(arg, cmdline) {
```

Comment:
please do no parse the kernel cmdline a second time. it's already parsed in parse_proc_cmdline_item(). please store any metainfo you need in struct crypt_device